### PR TITLE
WIP: Port bash script naveen521kk to .travis.yml

### PR DESCRIPTION
**Pull request by stuaxo**
Port WHL build script @naveen521kk to plain `.travis.yml`

- Delete build script and move to `.travis.yml`

Minor changes: 
- Introduce "ARCH" variable for windows, x86 or x64
- Rename PYVER to PYTHON_VERSION to be consistent with CAIRO_VERSION
- Downloads are grouped with downloads from other platforms (in theory we could fail early here if they fail)
- Files are downloaded to another folder (coverage was trying to check the python installation!)
- Added the newer tests from newer pycairo

TODO
Having added the newer tests, test_typing fails.

### CODE REVIEW

**Comment by naveen521kk** [comment]
**Question**: Any necessity for it to be merged in master? Because I used the branch `windows-wheels` for the other PR. Also, I allowed maintainer to commit directly in that branch. SO you could directly commit if needed. Or if you want me to review the code make a PR to `windows-wheels` branch.
Cheers

**Comment by stuaxo** [comment]
> **Question**: Any necessity for it to be merged in master? Because I used the branch `windows-wheels` for the other PR. Also, I allowed maintainer to commit directly in that branch. SO you could directly commit if needed. Or if you want me to review the code make a PR to `windows-wheels` branch.
> Cheers

Nope, this was definitely my error !   In a bit of a rush now :)  
It would be great if you can sanity check this... I think it works, but I haven't tested the whls for a couple of days (when I was using a modified version of your script).

**Comment by stuaxo** [comment]
I've got to be AFK for a while so can't wait to see tests pass, it's possible I've done something silly like bodged the merge with the latest version of your work.

**Comment by naveen521kk** [comment]
Currently I am busy I will look into this maybe by tomorrow.

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml]: This can be updated to `3.8.5` as it is released now.
```suggestion
      env: ARCH=x86 CAIRO_VERSION=1.17.2 PYTHON_PACKAGE=pythonx86 PYTHON_VERSION="3.8.5" 
```

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml]: This can be updated to `3.8.5` as it is released now.
```suggestion
      env: ARCH=x64 CAIRO_VERSION=1.17.2 PYTHON_PACKAGE=python PYTHON_VERSION="3.8.5"
```

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml:93]: What does the above two lines do? I really don't understand.

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml:81]: Why this is removed?

**Review by naveen521kk — CHANGES_REQUESTED** [review]
This was some quick questions I came up. Also, can we create a GitHub Action (As it is the only one I know has scheduled build) which checks for upstream for cairo version and other things? A python script would do. Just execute the py script and then create an issue if it not update. Any ideas about this?

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:81]: We def want this, I think it may have been failing as the token needs setting up at this end, so may need to be in another commit... double check though, as if it works already then it should stay.

File: .travis.yml
Comment (stuaxo) [reply .travis.yml]: Good spot...

We can automate this, I spoke to dependencies.io who gave me an account for open source stuff + that can generate diffs and PRs for this sort of thing. 

It's annoying nuget doesn't accept a query like 3.7.x and give the highest match, I wonder if chocolatey can do it?

Comment (naveen521kk) [reply .travis.yml]: You can use a query like `3.7` I think.

File: .travis.yml
Comment (naveen521kk) [reply .travis.yml:81]: The tokens only apply to commits in master branch or releases and not PR. I noticed it in many places. So this should be added. :)

**Comment by naveen521kk** [comment]
> Also, can we create a GitHub Action (As it is the only one I know has scheduled build) which checks for upstream for cairo version and other things? A python script would do. Just execute the py script and then create an issue if it not update. Any ideas about this?

What do you think about this? @stuaxo

File: .travis.yml
Comment (naveen521kk) [reply .travis.yml]: Also I just found that choco and nuget are one and the same thing in structure. Choco uses nugets structure.

**Comment by stuaxo** [comment]
I can def do the github action to generate patch PRs, I have one I need to push upstream to cairo-windows that already does this, so fairly straightforward.

I'm a bit new to github actions so wasn't sure the best events for it to run on, so am thinking scheduled daily, and also available from a webhook (I imagine only @lazka will have the token for that).

As soon as someone merges the PR then I guess that will trigger a build and release (on github at least).

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:81]: Will definitely re-add, TBH when I pasted that in was in a bit of a hurry :D

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml:76]: using a directory in `$TMP` variable is lot better instead of directly temp directory. What do you say?

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml]: This looks weird to me. Should it be.
```suggestion
  - if [ "$TRAVIS_OS_NAME" == "windows" ]; then export PATH=$TMP/nuget-$ARCH/python.$PYTHON_VERSION/tools/:$TMP/nuget-$ARCH/python.$PYTHON_VERSION/tools/Scripts:$PATH; fi
```

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml]: This line can be executed for all OS.
```suggestion
  - python -m pip install --upgrade wheel
```

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Some suggetsions

**Comment by naveen521kk** [comment]
About test failing I remember I read somewhere where lazka said about `tee surface` not implemented in windows or something like that. So its good to ping(@lazka) to help us on this or can we exempt these test on windows. @stuaxo

We can skip that test for windows see https://docs.pytest.org/en/latest/skipping.html#id1

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:81]: Re-added - slightly, naughtily ammended and I force pushed this,  to avoid there being a change in the history that removes it.

File: .travis.yml
Comment (stuaxo) [reply .travis.yml]: Chocolatey has some extensions, I'm inclined to switch back to it, as it has some advantages.
Native support in Travis: one less thing to install, and Travis' cache support seems to include it. 

You can force it to install for 32 or 64 bit as needed. 

Semver: not supported now, but they have a ticket.
https://github.com/chocolatey/choco/issues/1610
 
In the meantime I think we might be able to do semver support with a combination of powershell and the output of `choco search` or hit their API directly with `cURL`. 

This would be good as we could specify the python version with 2 digit versions, not 3, so one less thing to support and patch.

Comment (stuaxo) [reply .travis.yml]: > You can use a query like `3.7` I think.

This doesn't work directly right now, so you get `3.7.0` :( but yeah, we can defintely use a query like that... I'm having a play with the choco commandline to see if I can persaude it to list all versions, not much luck yet - but you can see on the site they exist, so there will be a way of doing it somehow.

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:93]: (handwaves) - checking how much coverage there is, I think it uploads to a service, where you can view it.

I haven't used it to much, the coverage bits where from newest pycairo master, not my own changes.

**Comment by stuaxo** [comment]
> 
> 
> > Also, can we create a GitHub Action (As it is the only one I know has scheduled build) which checks for upstream for cairo version and other things? A python script would do. Just execute the py script and then create an issue if it not update. Any ideas about this?
> 
> What do you think about this? @stuaxo

Definitely doable, I think I wrote somewhere else we can do with dependencies.io and their deps tool.

I wonder which things we want to check on... if we can get rid of the need for an exact python build. then that leaves `cairo-windows`, at least until we make whls for other platforms.

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:76]: Yes... *but* lets try and switch back to chocolatey and see if we can avoid this whole bit.

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:92]: So.. .discovered some things.   There is [a bug](https://github.com/chocolatey/choco/issues/1843) in current chocolatey (fixed in next version) that prevents this kind of search.

But in the meantime it's possible to downgrade chocolatey:

`choco install chocolatey -y --version 0.10.13 --allow-downgrade`

Then we can search all versions:

```shell
$ choco search -e python3 -a
Chocolatey v0.10.13
python3 3.9.0-rc1 [Approved]
python3 3.9.0-b5 [Approved]
python3 3.9.0-b4 [Approved]
[snipped a lot of versions]
```

Use grep to search for python 3.7 and remove the [Approved] column:
```shell
$ choco search -e python3 -a | grep 'python3 3.7' | head | cut -d' ' -f1,2
python3 3.7.9
python3 3.7.8
python3 3.7.7
python3 3.7.6.20200110
python3 3.7.6
python3 3.7.5
python3 3.7.4
python3 3.7.3
python3 3.7.2
python3 3.7.2-rc1
```

^ It looks like they come back in the order they were added, so which seems like what we want.

Get exact python version given "3.7":
```
choco search -e python3 -a | grep 'python3 3.7' | head | cut -d' ' -f2 | head -1
3.7.9
```

File: .travis.yml
Comment (stuaxo) [reply .travis.yml]: So.. .discovered some things.   There is [a bug](https://github.com/chocolatey/choco/issues/1843) in current chocolatey (fixed in next version) that prevents this kind of search.

But in the meantime it's possible to downgrade chocolatey:

`choco install chocolatey -y --version 0.10.13 --allow-downgrade`

Then we can search all versions:

```shell
$ choco search -e python3 -a
Chocolatey v0.10.13
python3 3.9.0-rc1 [Approved]
python3 3.9.0-b5 [Approved]
python3 3.9.0-b4 [Approved]
[snipped a lot of versions]
```

Use grep to search for python 3.7 and remove the [Approved] column:
```shell
$ choco search -e python3 -a | grep 'python3 3.7' | head | cut -d' ' -f1,2
python3 3.7.9
python3 3.7.8
python3 3.7.7
python3 3.7.6.20200110
python3 3.7.6
python3 3.7.5
python3 3.7.4
python3 3.7.3
python3 3.7.2
python3 3.7.2-rc1
```

^ It looks like they come back in the order they were added, so which seems like what we want.

Get exact python version given "3.7":
```
choco search -e python3 -a | grep 'python3 3.7' | head | cut -d' ' -f2 | head -1
3.7.9
```

Comment (naveen521kk) [reply .travis.yml]: I really don't know what to tell on this. But yeah I prefer using choco but does it has x86 python? I could find one so I use nuget for it and was also in [python docs](https://docs.python.org/3/using/windows.html#windows-nuget). Think about it before using. Also, we can use nuget feed in choco by

```powershell
choco install python -source="https://www.nuget.org/api/v2"
```
This will `nuget`  feed instead of chocolatey's.

**Comment by naveen521kk** [comment]
How about py `3.9` binary?

File: .travis.yml
Comment (stuaxo) [reply .travis.yml]: We can pass  - -x86 to install 32 bit versions, I haven't tried it yet.

Comment (naveen521kk) [reply .travis.yml]: That doesn't help much. Nuget python is build for CI ,but choco is using default py installer which may be troublesome. So better to use nuget for x86.

Comment (stuaxo) [reply .travis.yml]: Fair enough, will stay on nuget, if there are questions around chocolatey.

Do you know what the issues with the default installer are?

I haven't used windows as my default OS for a while, well aware windows + python + C extensions can get pretty painful.

Comment (naveen521kk) [reply .travis.yml]: I kinda use windows as my default one. Maybe because I don't know C or something. I just know quite a few basic of python and got to know about this because of [manim](https://GitHub.com/manimcommunity/manim). Some video tool which use pycairo as it's base. I don't even have C compilers in my PC and that made me do this. But yeah Nuget x86 py build should be used or we should use something kinada docker or something but not required.

**Comment by stuaxo** [comment]
We're OK until October the 5th for 3.9
https://www.python.org/dev/peps/pep-0596/#schedule

File: .travis.yml
Comment (stuaxo) [reply .travis.yml:76]: Oh, DYM set something like PYCAIRO_TMP=$TMP/pycairo and use that ?

It's definitely doable, I was trying to balance adding more lines and variables to remember, but it could clean things up.

**Comment by naveen521kk** [comment]
I will need to check and work something and will do in two days and later it can be merged.

**Comment by naveen521kk** [comment]
Now tests should be passing.

**Comment by naveen521kk** [comment]
@stuaxo We messed up something with x86 builds.

**Comment by naveen521kk** [comment]
@stuaxo If the tests pass this is ready to be merged.

**Review by stuaxo — COMMENTED** [review]
I think we can change the test to check for minimum version to also check for Windows...

Do we know everything works OK with TeeSurface not enabled?

We might be able to just enable TeeSurface upstream, I don't have the dev environment setup to try this right now, but if it works we could always patch upstream and this will be fixed for good (it seems like msys has this enabled).

**Comment by naveen521kk** [comment]
> it seems like msys has this enabled

I noticed about that. Maybe let's use that? Include the `cairo.dll` it generates to the `.whl` files. That would do. Also, I have tested this with the site which provided wheel that TeeSurface doesn't work on it also. Also, maybe it would be better to chat on gitter that these comments.

**Comment by naveen521kk** [comment]
> Do we know everything works OK with TeeSurface not enabled?

I am such a noob here. I don't know C lol. So, I have no much experience in pycairo also.

**Comment by naveen521kk** [comment]
I got some weird ideas. @stuaxo 
- Why not move from Travis to GitHub Actions because Travis kinda sucks me ( I have more experience with Actions). 
- Not to depend on preshing repo instead use msys for our needs( There a gh-actions for that which makes it simple than Travis).

**Comment by naveen521kk** [comment]
Looks like I won't be able to spend much time on this in the upcoming week and I plan to take a long break. So I would be working on this PR next week.
Just tell me what to do it would be good. If I couldn't complete it then you(or anyone else) should be taking care of this 🙂. Also, I am adding you as collaborators to this fork. Please don't delete any release as some project depend on them.
Yes, I am running out of time and this would be the last week. Next, I won't be available. 
CC
@stuaxo

**Comment by stuaxo** [comment]
I've got a similar time problem..  Will reply to this shortly though :)

**Comment by stuaxo** [comment]
> I got some weird ideas. @stuaxo
> 
>     * Why not move from Travis to GitHub Actions because Travis kinda sucks me ( I have more experience with Actions).
> 
Yeah, definitely if @lazka is up for it.

>     * Not to depend on preshing repo instead use msys for our needs( There a gh-actions for that which makes it simple than Travis).
This makes sense, but I reckon we can probably be able to land this first, as it does seem to work.

We should verify there are no new issues using this when pygobject is already installed, then it should be good to go.

I'm also going to have less time, but should be able to help land the WHL support at least.

**Comment by naveen521kk** [comment]
**Question**: Does the whl available on https://www.lfd.uci.edu/~gohlke/pythonlibs/#pycairo enables tee surface? We need to look into it.

**Comment by naveen521kk** [comment]
> Does the whl available on https://www.lfd.uci.edu/~gohlke/pythonlibs/#pycairo enables tee surface? 

Yes it works there.

**Comment by naveen521kk** [comment]
@stuaxo If you could https://github.com/naveen521kk/pycairo/runs/1035437105?check_suite_focus=true#step:6:165 help me fix this I would be very happy and also it can be used by preshing/cairo-windows and use automatic releasing using Github actions. After that, maybe lets depend on then.

File: .travis.yml
Comment (naveen521kk) [reply .travis.yml]: This can be easily fixed if we were to use GitHub Actions.

**Comment by stuaxo** [comment]
I'll have a look, I'm back @ work at the moment so won't be until later.

I've been looking at getting TeeSurface enabled on cairo-windows, which would mean we don't nee to change any tests
https://github.com/preshing/cairo-windows/issues/7

@preshing has suggested a solution to get it working, but I won't have time until @ least later on,
If you have a VS2017 build environment it might be worth trying.


I started a PR there to generate make PRs when there are new releases of cairo, freetype etc update using dependencies.io
https://github.com/preshing/cairo-windows/pull/10
I probably need to add some instructions on what setup is needed so it's easier for @preshing to land it.

**Comment by stuaxo** [comment]
Quick feedback:  I can see it's building with VS2019, I don't really know the differences between 2017 + 2019, but @preshing told me that cairo-windows is currently built on 2017, not 19, so there may be some compatibility issue.

**Comment by naveen521kk** [comment]
Yes @stuaxo does https://github.com/naveen521kk/pycairo/runs/1046803166?check_suite_focus=true work? I am asking on tee surface did that work? If you any test environment download the wheels from artifact's and test it once locally. It that works we can tell @preshing to release it so. And yes tee surface problem will be over. 
**Edit**: I have done something stupid for x86 builds. So please ignore that one instead use the one from the latest commit.

**Comment by stuaxo** [comment]
I'm assuming this will pass now @naveen521kk are you up for merging this into your branch so there's one less to consider?

**Comment by stuaxo** [comment]
@naveen521kk I'm away from my computer, any idea why the x64 bit build fails?

All I tried was telling Travis to rebuild from my phone, think I need to verify from windows VM when in front of computer later.

File: .travis.yml
Comment (naveen521kk) [inline .travis.yml]: I found why it fails. Because the nuget package `pythonx86` contains x86 build of python and it does pull x64 python and then later tries to build for x86 which makes is fail. :( Check my shell script.

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Here is what happens.

**Comment by stuaxo** [comment]
I need to stop trying to edit this at the same time as trying to get a 3 year old ready to get out of the house :)   I'll come back to this when I can give it my full attention and stop adding duff commits :)

Don't worry, I will squash the commits before we push anything to the pycairo proper :)

**Comment by stuaxo** [comment]
> I need to stop trying to edit this at the same time as trying to get a 3 year old ready to get out of the house :) I'll come back to this when I can give it my full attention and stop adding duff commits :)
> 
> Don't worry, I will squash the commits before we push anything to the pycairo proper :)

There's a pretty good chance it might be the when the work week starts (Tuesday it's a long weekend) when I can look at stuff properly, as it's almost all small person time, then just tiny bits of adult time during the weekend (and some small bits of coding time)!

**Comment by naveen521kk** [comment]
Maybe I will work if I get some time.

**Comment by naveen521kk** [comment]
Any idea why https://travis-ci.com/github/naveen521kk/pycairo/jobs/379364608 fail? @stuaxo 
Gonna rerun it.

**Comment by stuaxo** [comment]
Squashed and merged to pygobject/pycairo.

**Pull request closed without merging**
